contracts/PooledStaking.sol

```diff
@@ -0,0 +1,512 @@
+      );
+
+      if (current.next != 0) {
+        Deallocation storage next = deallocations[current.next];
+        require(
+          // require its deallocation time to be greater than new deallocation time
+          next.deallocateAt > deallocateAt,
+          "Deallocation time must be smaller than next deallocation"
+        );
+      }
+
+      // get next available id
+      uint id = deallocationCount;
+      deallocationCount++;
+
+      // point to our new deallocation
+      uint next = current.next;
+      current.next = id;
+
+      deallocations[id] = Deallocation(
+        requested, deallocateAt, contractAddress, msg.sender, next
+      );
+
+      // increase pending deallocation amount so we keep track of final allocation
+      uint newPending = staker.pendingDeallocations[contractAddress].add(requested);
+      staker.pendingDeallocations[contractAddress] = newPending;
+    }
+  }
+
+  function pushBurn(address contractAddress, uint amount) onlyInternal external {
+
+    burns[burnCount] = Burn(amount, now, contractAddress, 0); // add new burn
+
+    // TODO: recheck this
+    // burnCount and firstBurn can be equal only when they're both 0
+    // i.e. when pushing the first burn ever
+    // if those are different, burnCount - 1 will be >= 0 (will not underflow)
+    // but we should check that it doesn't point to an empty slot (processed & deleted burn)
+    if (burnCount != firstBurn && burns[burnCount - 1].burnedAt > 0) {
+      burns[burnCount - 1].next = burnCount; // set previous burn's next to current burn id
+    } else {
+      // otherwise this is the only existing burn, therefore it is the first one as well
+      firstBurn = burnCount;
+    }
+
+    ++burnCount; // update counter
+  }
+
+  function pushReward(address contractAddress, uint amount, address from) onlyInternal external {
+
+    // transfer tokens from specified contract to us
+    // token transfer should be approved by the specified address
+    Vault.deposit(token, from, amount);
+
+    rewards[rewardCount] = Reward(amount, now, contractAddress, 0); // add new reward
+
+    // TODO: recheck this
+    // similar to pushBurn
+    if (rewardCount != firstReward && rewards[rewardCount - 1].rewardedAt > 0) {
+      rewards[rewardCount - 1].next = rewardCount; // set previous reward's next to current id
+    } else {
+      firstReward = rewardCount;
+    }
+
+    ++rewardCount; // update counter
+  }
+
+  function processPendingActions() public {
+    while (true) {
+      Burn storage burn = burns[firstBurn];
+      Deallocation storage deallocation = deallocations[firstDeallocation];
+
+      bool canBurn = burn.burnedAt != 0; // end not reached
+      bool canDeallocate = deallocation.deallocateAt != 0 && deallocation.deallocateAt <= now;
+
+      if (!canBurn && !canDeallocate) {
+        // everything is processed
+        break;
+      }
+
+      // TODO: check if there's enough gas for the next iteration
```

Comment:
the gas check here would now on _contract.stakers.length * staker.contracts.length so we'd need some bounds there to know if there is enough gas